src/security/collection.cpp

```diff
@@ -0,0 +1,396 @@
+    dnsec::ddwaf_obj *root = memres_.allocate_objects<dnsec::ddwaf_obj>(1);
+    dnsec::ddwaf_map_obj &root_map = root->make_map(2, memres_);
+
+    set_response_status(request, root_map.at_unchecked(0));
+    set_response_headers_no_cookies(request, root_map.at_unchecked(1));
+
+    return root;
+  }
+
+ private:
+  static void set_map_entry_str(dnsec::ddwaf_obj &slot, std::string_view key,
+                                const ngx_str_t &value) {
+    slot.set_key(key);
+    slot.make_string(to_string_view(value));
+  }
+
+  void set_request_query(const ngx_http_request_t &request,
+                         dnsec::ddwaf_obj &slot) {
+    slot.set_key(kQuery);
+    const ngx_str_t &query = request.args;
+    if (query.len == 0) {
+      slot.make_array(nullptr, 0);
+      return;
+    }
+
+    dnsec::QueryStringIter it{query, memres_, '&',
+                              dnsec::QueryStringIter::trim_mode::no_trim};
+    set_value_from_iter(it, slot);
+  }
+
+  template <typename Iter>
+  void set_value_from_iter(Iter &it, dnsec::ddwaf_obj &slot) {
+    // first, count the number of occurrences for each key
+    std::unordered_map<std::string_view, std::size_t> keys_bag;
+    for (; !it.ended(); ++it) {
+      std::string_view key = it.cur_key();
+      // if (it.is_delete()) { }
+      // don't reduce the count, we need to allocate space for the provisional
+      // writes (before the deletion)
+      // we could improve this by saving the pointer of the first non-deleted
+      // header per key
+      keys_bag[key]++;
+    }
+
+    // we now know the number of keys; allocate map entries
+    dnsec::ddwaf_obj *entries =
+        memres_.allocate_objects<dnsec::ddwaf_obj>(keys_bag.size());
+    slot.make_map(entries, keys_bag.size());
+    dnsec::ddwaf_obj *next_free_entry = entries;
+
+    // fill the map entries
+    // map that saves the ddwaf_object for keys that occurr more than once
+    std::unordered_map<std::string_view, dnsec::ddwaf_arr_obj *>
+        indexed_entries;
+    for (it.reset(); !it.ended(); ++it) {
+      auto [key, value] = *it;
+      std::size_t const num_occurr = keys_bag[key];
+
+      // common scenario: only 1 occurrence of the key
+      if (num_occurr == 1) {
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+        entry.set_key(key);
+        entry.make_string(value);
+        continue;
+      }
+
+      auto ie = indexed_entries.find(key);
+      if (ie == indexed_entries.end()) {
+        // first occurrence of this key
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+
+        entry.set_key(key);
+        auto &arr_val = entry.make_array(num_occurr, memres_);
+        indexed_entries.insert({key, &arr_val});
+
+        if (!it.is_delete()) {
+          arr_val.at_unchecked<dnsec::ddwaf_obj>(0).make_string(value);
+          entry.nbEntries = 1;
+        }
+      } else {
+        // subsequent occurrence of this key
+        auto &arr_val = *ie->second;
+        if (!it.is_delete()) {
```

Comment:
What does this mean? When would `is_delete` be true?

Reply:
modules can delete existing headers. For this, they add a header with hash = 0 (and add headers with hash 1; hence "hash" doesn't represent a hash in response headers like it does in request headers, even though they use the same data structure)

Reply:
Ah that makes sense, thanks for explaining.